tokio/tests/sync_set_once.rs

```diff
@@ -0,0 +1,151 @@
+#![warn(rust_2018_idioms)]
+#![cfg(feature = "full")]
+
+use std::mem;
+use std::sync::{
+    atomic::{AtomicU32, Ordering},
+    Arc,
+};
+use tokio::sync::SetOnce;
+
+#[derive(Clone)]
+struct DropCounter {
+    drops: Arc<AtomicU32>,
+}
+
+impl DropCounter {
+    fn new() -> Self {
+        DropCounter {
+            drops: Arc::new(AtomicU32::new(0)),
+        }
+    }
+
+    fn assert_num_drops(&self, value: u32) {
+        assert_eq!(value, self.drops.load(Ordering::Relaxed));
+    }
+}
+
+impl Drop for DropCounter {
+    fn drop(&mut self) {
+        self.drops.fetch_add(1, Ordering::Relaxed);
+    }
+}
+
+#[test]
+fn drop_cell() {
+    let fooer = DropCounter::new();
+    let fooer_cl = fooer.clone();
+
+    {
+        let once_cell = SetOnce::new();
+        let prev = once_cell.set(fooer_cl);
+        assert!(prev.is_ok())
+    }
+
+    fooer.assert_num_drops(1);
+}
+
+#[test]
+fn drop_cell_new_with() {
+    let fooer = DropCounter::new();
+
+    {
+        let once_cell = SetOnce::new_with(Some(fooer.clone()));
+        assert!(once_cell.initialized());
+    }
+
+    fooer.assert_num_drops(1);
+}
+
+#[test]
+fn drop_into_inner() {
+    let fooer = DropCounter::new();
+
+    let once_cell = SetOnce::new();
+    assert!(once_cell.set(fooer.clone()).is_ok());
+    let val = once_cell.into_inner();
+    fooer.assert_num_drops(0);
+    drop(val);
+    fooer.assert_num_drops(1);
+}
+
+#[test]
+fn drop_into_inner_new_with() {
+    let fooer = DropCounter::new();
+
+    let once_cell = SetOnce::new_with(Some(fooer.clone()));
+    let val = once_cell.into_inner();
+    fooer.assert_num_drops(0);
+    mem::drop(val);
+    fooer.assert_num_drops(1);
+}
+
+#[test]
+fn from() {
+    let cell = SetOnce::from(2);
+    assert_eq!(*cell.get().unwrap(), 2);
+}
+
+#[test]
+fn set_and_get() {
+    static ONCE: SetOnce<u32> = SetOnce::const_new();
+
+    ONCE.set(5).unwrap();
+    let value = ONCE.get().unwrap();
+    assert_eq!(*value, 5);
+}
+
+#[tokio::test]
+async fn set_and_wait() {
+    static ONCE: SetOnce<u32> = SetOnce::const_new();
+
+    let _ = tokio::spawn(async { ONCE.set(5) }).await;
+    let value = ONCE.wait().await;
+    assert_eq!(*value, 5);
+}
```

Comment:
Can we also have a version of this test using `std::thread::spawn`?

And also a test that spawns two real threads and calls `set` from both, asserting that exactly one succeeds. Then we have tests that correspond to both loom tests.